crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2484,17 +2509,32 @@ impl<'db> TypeInferenceBuilder<'db> {
 
         let left_ty = self.infer_expression(left);
         let right_ty = self.infer_expression(right);
+        self.infer_binary_expression_type(binary, *op, left_ty, right_ty)
+    }
 
+    fn infer_binary_expression_type(
+        &mut self,
+        expr: &ast::ExprBinOp,
+        op: ast::Operator,
+        left_ty: Type<'db>,
+        right_ty: Type<'db>,
+    ) -> Type<'db> {
         // Check for division by zero; this doesn't change the inferred type for the expression, but
         // may emit a diagnostic
         if matches!(
             (op, right_ty),
             (
                 ast::Operator::Div | ast::Operator::FloorDiv | ast::Operator::Mod,
-                Type::IntLiteral(0),
+                Type::IntLiteral(0)
             )
         ) {
-            self.check_division_by_zero(binary, left_ty);
+            // We want to print the original type that was used in the expression. The left type might
```

Comment:
I would prefer to leave this check out in `infer_binary_expression` and not re-check it recursively, and then we don't have to bother with `original_type` and this long comment. The only cost is that we have to explicitly check for `BooleanLiteral(False)` in the divisor, but that seems totally fine. It doesn't seem worth the complexity of adding `original_type` and needing this long comment, and the efficiency cost of double-checking for division by zero, just so we can match on `IntLiteral(0)` instead of `IntLiteral(0) | BooleanLiteral(False)`.

(It would be different if there were lots of different equivalent-to-zero values we might be converting to in a recursive call, but there aren't -- there's exactly one.)